### PR TITLE
Make NME and NGE math blocks agnostic to order of operations and left vs. right input

### PR DESCRIPTION
**Pull request by ryantrem**
Trying out NGE, I found it a bit confusing that the left and right inputs of math nodes were treated differently and that the order of operations mattered. For example:

NGE-Connection

In this change, I made it so that it is equivalent whether an output of another node is connected to the left or right input of a math node, and also the order of operations doesn't matter. Each time the connection state changes, it re-evaluates the accepted types for each input (left and right), as well as the output type based on those input types.

NGE-Flexible-Multiply

~~I realized part way through doing this that mathBlock is NGE specific, and the set of math related blocks for NME have the same limitations, so ideally these changes would be made for both (or the same mathBlock would be shared between the two if possible).~~
EDIT: There are differences between the NME and NGE math related blocks, so while I didn't end up sharing the code between the two, I did factor out a `MathBlock` base class for the NME Add/Subtract/Multiply/Divide blocks that has logic similar to what I added to the NGE `MathBlock`.

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/14857/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14857/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14857/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/14857/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/14857/merge
https://gui.babylonjs.com/?snapshot=refs/pull/14857/merge
https://nme.babylonjs.com/?snapshot=refs/pull/14857/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/14857/merge#BCU1XR#0

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14857/merge/testResults/webgl2playwright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14857/merge/testResults/webgpuplaywright/index.html

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14857/merge/testResults/webgpuplaywright/index.html

File: packages/dev/core/src/Meshes/Node/Blocks/mathBlock.ts
Comment (RaananW) [inline packages/dev/core/src/Meshes/Node/Blocks/mathBlock.ts:8]:
```suggestion
import type { Observer } from "core/Misc/observable";
```

**Review by Popov72 — APPROVED** [review]

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/14857/merge/testResults/webgpuplaywright/index.html

**Review by Popov72 — APPROVED** [review]

**Review by sebavan — APPROVED** [review]

**Comment by ryantrem** [comment]
Since I have very little prior experience using NGE/NME and this is the first time I've touched the code, there is some newbie risk to this change. May be good to get a once over from @deltakosh, and we could merge it now, or after the 7.0 release.

File: packages/dev/core/src/Materials/Node/Blocks/mathBlock.ts
Comment (deltakosh) [inline packages/dev/core/src/Materials/Node/Blocks/mathBlock.ts:106]: I guess you also need to clear the this._connectionObservers array

**Review by deltakosh — APPROVED** [review]
Just a tiny question
Love it!

File: packages/dev/core/src/Materials/Node/Blocks/mathBlock.ts
Comment (ryantrem) [reply packages/dev/core/src/Materials/Node/Blocks/mathBlock.ts:106]: `_connectionObservers` is intended to essentially be a tuple of 4 observers, and to be the equivalent of having 4 properties on the class. It's just that storing them as an array to slightly simplify the logic. Clearing the "array" would be equivalent to setting 4 separate properties to null/undefined, which seems a like maybe dispose overkill. I could do it, but to do so I need to make the code more permissive (e.g. mutable array). Personally I think being more restrictive is more valuable because more restrictive code has fewer opportunities for bugs, but I don't feel super strongly about it. If you feel more strongly about clearing the array, I can either make it not readonly, or split it out into 4 properties so there is no array to be cleared, just let me know!

**Comment by ryantrem** [comment]
I think there is a bug with part of this change, pls don't merge yet. :)

File: packages/dev/core/src/Materials/Node/Blocks/mathBlock.ts
Comment (deltakosh) [reply packages/dev/core/src/Materials/Node/Blocks/mathBlock.ts:106]: The problem is GC, If someone holds a link to a mathBlock, if you dispose the block there is still a link between the this._connectionObservers and the observer who in turn are keeping link to their "Action"

Or do I miss something?

Comment (ryantrem) [reply packages/dev/core/src/Materials/Node/Blocks/mathBlock.ts:106]: The observers are all state contained within the mathBlock itself (not external), so as it is that state/memory stays alive with the mathBlock (same as values of other properties of the class, for example). Clearing this collection would allow some of the memory that makes up the class be GC'd early if you still held a reference to the mathBlock. That could be an optimization, and in my mind that would be the same optimization as if we said we should always set all properties of a class (that reference any kind of object) to null upon dispose, which I don't think we do elsewhere. For example, nodeMaterialBlock does not set _buildTarget to null in dispose. Similarly, if instead of holding these observers in an array I held them in individual properties (e.g. leftConnectionObservable, leftDisconnectionObservable, rightConnectionObservable, rightDisconnectionObservable), I wouldn't expect the dispose function to set all those properties to null (the equivalent of clearing the collection).

Anyway, just providing that as explanation of my thinking. I'm not terribly opposed to changing it, it just means I need to remove the readonly-ness of the collection which obscures the intent of the code a bit.

Comment (deltakosh) [reply packages/dev/core/src/Materials/Node/Blocks/mathBlock.ts:106]: I disagree :)
Every observable are cleared on dispose because they can hold external references
https://github.com/BabylonJS/Babylon.js/blob/master/packages/dev/core/src/scene.ts#L4905

And you don't need to remove readOnly (i think) if you set array.length = 0

Comment (ryantrem) [reply packages/dev/core/src/Materials/Node/Blocks/mathBlock.ts:106]: Also, it might be more clear that I'm basically just mashing 4 properties into 1 if I changed the type to:
`private readonly _connectionObservers: Tuple<Observer<NodeMaterialConnectionPoint>, 4>;`

**Pull request merged**
